precise/iaf_psc_exp_ps_lossless.cpp

```diff
@@ -0,0 +1,671 @@
+  V_.expm1_tau_in_ = numerics::expm1( -V_.h_ms_ / P_.tau_in_ );
+  V_.P20_ = -P_.tau_m_ / P_.c_m_ * V_.expm1_tau_m_;
+  V_.P21_ex_ = -P_.tau_m_ * P_.tau_ex_ / ( P_.tau_m_ - P_.tau_ex_ ) / P_.c_m_
+    * ( V_.expm1_tau_ex_ - V_.expm1_tau_m_ );
+  V_.P21_in_ = -P_.tau_m_ * P_.tau_in_ / ( P_.tau_m_ - P_.tau_in_ ) / P_.c_m_
+    * ( V_.expm1_tau_in_ - V_.expm1_tau_m_ );
+  V_.refractory_steps_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  assert( V_.refractory_steps_
+    >= 0 ); // since t_ref_ >= 0, this can only fail in error
+
+  V_.a1_ = P_.tau_m_ * P_.tau_ex_;
+  V_.a2_ = P_.tau_m_ * ( P_.tau_m_ - P_.tau_ex_ );
+  V_.a3_ = P_.c_m_ * P_.U_th_ * ( P_.tau_m_ - P_.tau_ex_ );
+  V_.a4_ = P_.c_m_ * ( P_.tau_m_ - P_.tau_ex_ );
+
+  V_.b1_ = -P_.tau_m_ * P_.tau_m_;
+  V_.b2_ = P_.tau_m_ * P_.tau_ex_;
+  V_.b3_ = P_.tau_m_ * P_.c_m_ * P_.U_th_;
+  V_.b4_ = -P_.c_m_ * ( P_.tau_m_ - P_.tau_ex_ );
+
+  V_.c1_ = P_.tau_m_ / P_.c_m_;
+  V_.c2_ =
+    ( -P_.tau_m_ * P_.tau_ex_ ) / ( P_.c_m_ * ( P_.tau_m_ - P_.tau_ex_ ) );
+  V_.c3_ = ( P_.tau_m_ * P_.tau_m_ ) / ( P_.c_m_ * ( P_.tau_m_ - P_.tau_ex_ ) );
+  V_.c4_ = P_.tau_ex_ / P_.tau_m_;
+  V_.c5_ = ( P_.c_m_ * P_.U_th_ ) / P_.tau_m_;
+  V_.c6_ = 1 - ( P_.tau_ex_ / P_.tau_m_ );
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::iaf_psc_exp_ps_lossless::update( const Time& origin,
+  const long from,
+  const long to )
+{
+  assert( to >= 0 );
+  assert( static_cast< delay >( from )
+    < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  // at start of slice, tell input queue to prepare for delivery
+  if ( from == 0 )
+  {
+    B_.events_.prepare_delivery();
+  }
+
+  /* Neurons may have been initialized to superthreshold potentials.
+     We need to check for this here and issue spikes at the beginning of
+     the interval.
+  */
+  if ( S_.y2_ >= P_.U_th_ )
+  {
+    emit_instant_spike_( origin,
+      from,
+      V_.h_ms_ * ( 1.0 - std::numeric_limits< double >::epsilon() ) );
+  }
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+    // time at start of update step
+    const long T = origin.get_steps() + lag;
+
+    // if neuron returns from refractoriness during this step, place
+    // pseudo-event in queue to mark end of refractory period
+    if ( S_.is_refractory_
+      && ( T + 1 - S_.last_spike_step_ == V_.refractory_steps_ ) )
+    {
+      B_.events_.add_refractory( T, S_.last_spike_offset_ );
+    }
+
+    // save state at beginning of interval for spike-time approximation
+    V_.y0_before_ = S_.y0_;
+    V_.I_syn_ex_before_ = S_.I_syn_ex_;
+    V_.I_syn_in_before_ = S_.I_syn_in_;
+    V_.y2_before_ = S_.y2_;
+
+    // get first event
+    double ev_offset;
+    double ev_weight;
+    bool end_of_refract;
+
+    if ( not B_.events_.get_next_spike(
+           T, true, ev_offset, ev_weight, end_of_refract ) )
```

Comment:
This adds all the simultaneous spikes before a potential threshold crossing, i.e. a reset, meaning that the effect of all spikes that arrived in addition to those sufficient to trigger threshold crossing will be "lost", compared to the "standard" implementation of ``iaf_psc_exp_ps``.

I have not strong preference (at least I did not think long enough about this to have one), I'm just pointing out that this differs from the other models.

Reply:
I will need to look into this and its history.

Reply:
This is related to #368. In a [comment there](https://github.com/nest/nest-simulator/issues/368#issuecomment-222008250), I actually proposed that one should remove the accumulation altogether. It would simplify code and ensure consistency (in the way discussed in #368) for all precise models. I will work on it.